feat: wasm

richard-ramos opened the pull request:
How to compile zerokit for wasm and see example code:
1. Make sure you have nodejs installed and the `build-essential` package if using ubuntu.
2. Install wasm-pack
```
curl https://rustwasm.github.io/wasm-pack/installer/init.sh -sSf | sh
```
3. Compile zerokit for `wasm32-unknown-unknown`:
```
cd rln
wasm-pack build --release
```
4. Launch example app
```
cd www
npm install
npm start
```
5. Browse http://localhost:8080 and open the developer tools to see console logs

richard-ramos commented on rln/Cargo.toml:
```diff
@@ -6,18 +6,24 @@ edition = "2021"
-ark-ff = { version = "0.3.0", default-features = false, features = ["parallel", "asm"] }
-ark-std = { version = "0.3.0", default-features = false, features = ["parallel"] }
+ark-ec = { version = "0.3.0", default-features = false }
+ark-ff = { version = "0.3.0", default-features = false, features = [ "asm"] }
+ark-std = { version = "0.3.0", default-features = false }
 ark-bn254 = { version = "0.3.0" }
-ark-groth16 = { git = "https://github.com/arkworks-rs/groth16", rev = "765817f", features = ["parallel"] }
+ark-groth16 = { git = "https://github.com/arkworks-rs/groth16", rev = "765817f", default-features = false }
```
I removed the `parallel` feature to unsuccesfully avoid usage of threads

richard-ramos commented on rln/Cargo.toml:
```diff
@@ -6,18 +6,24 @@ edition = "2021"
 ark-serialize = { version = "0.3.0", default-features = false }
-ark-circom = { git = "https://github.com/gakonst/ark-circom", features = ["circom-2"] }
-wasmer = { version = "2.0" }
+ark-circom = { git = "https://github.com/vacp2p/ark-circom", branch = "wasm", features = ["circom-2"] }
```
This is a fork I created so (most of) this dependency can be run in wasm

richard-ramos commented on rln/Cargo.toml:
```diff
@@ -35,6 +41,17 @@ tiny-keccak = "2.0.2"
 # serialization
 serde = { version = "1.0.103", default-features = false, features = ["derive"] }
 serde_json = "1.0.48"
+base64 = "0.13.0"
```
This was added to make it easier to print logs for binary data. Will be removed before the PR is completed

richard-ramos commented on rln/Cargo.toml:
```diff
@@ -35,6 +41,17 @@ tiny-keccak = "2.0.2"
+console_error_panic_hook = "0.1.7"
+
+
+[profile.release]
+debug = true
```
This will be removed too

richard-ramos commented on rln/www/index.js:
```diff
@@ -0,0 +1,108 @@
+
+rln.init_panic_hook();
+
+function _base64ToArrayBuffer(base64) {
```
This helper function is for converting the circom, zkey and vk files from base64 to an `Uint8Array`

richard-ramos commented on rln/src/wasm.rs:
```diff
@@ -0,0 +1,108 @@
+#[allow(clippy::not_unsafe_ptr_arg_deref)]
+#[no_mangle]
+#[wasm_bindgen(js_name = RLNWitnessToJson)]
+pub fn rln_witness_to_json(ctx: *mut RLN, serialized_witness: Uint8Array) -> Object {
```
This function is used to obtain the input JSON required for the circom JS witness calculator

richard-ramos commented on rln/src/wasm.rs:
```diff
@@ -0,0 +1,108 @@
+#[allow(clippy::not_unsafe_ptr_arg_deref)]
+#[no_mangle]
+#[wasm_bindgen]
+pub fn generate_rln_proof_with_witness(
```
The difference between this function and what zerokit has to generate the proof is that it receives the witness calculator that was calculated on JS outside of zerokit/ark-circom

richard-ramos commented on rln/src/wasm.rs:
```diff
@@ -0,0 +1,108 @@
+    let rln = unsafe { &mut *ctx };
+
+    //    web_sys::console::log_1(&identitySecret.into());
+    let witness_vec: Vec<BigInt> = calculated_witness
+        .iter()
+        .map(|v| {
+            v.to_string(10)
+                .unwrap()
+                .as_string()
+                .unwrap()
+                .parse::<BigInt>()
+                .unwrap()
+        })
+        .collect();
```
I'm converting here the JS `BigInt` into a `numb_bigint::BigInt`. We can probably avoid this conversion if we somehow manage to use this datatype in JS

richard-ramos commented on rln/src/wasm.rs:
```diff
@@ -0,0 +1,108 @@
+    if rln.set_next_leaf(&input.to_vec()[..]).is_ok() {
+        Ok(())
+    } else {
+        Err("could not insert member into merkle tree".into())
```
`Err`s are throw in js and captured with try/catch

richard-ramos commented on rln/src/utils.rs:
```diff
@@ -85,7 +85,7 @@ pub fn fr_to_bytes_be(input: &Fr) -> Vec<u8> {
     let mut bytes: Vec<u8> = Vec::new();
     //We store the vector length
-    bytes.extend(input.len().to_le_bytes().to_vec());
+    bytes.extend(u64::try_from(input.len()).unwrap().to_le_bytes().to_vec());
```
This was added because `len()` returns an `usize`, so the number of bytes returned by `to_le_bytes` will change depending on the architecture

richard-ramos commented on rln/src/utils.rs:
```diff
@@ -121,7 +121,7 @@ pub fn vec_u8_to_bytes_be(input: Vec<u8>) -> Vec<u8> {
     let mut read: usize = 0;
 
-    let len = usize::try_from(u64::from_le_bytes(input[0..8].try_into().unwrap())).unwrap();
+    let len = u64::from_le_bytes(input[0..8].try_into().unwrap()) as usize;
```
This conversion should be safe, especially if we change lengths from u64 to u32. (See https://github.com/vacp2p/zerokit/issues/39)

richard-ramos commented on rln/src/ffi.rs:
```diff
@@ -42,8 +42,8 @@ impl<'a> From<&Buffer> for &'a [u8] {
     let input_data = <&[u8]>::from(unsafe { &*input_buffer });
-    let rln = RLN::new(tree_height, input_data);
-    unsafe { *ctx = Box::into_raw(Box::new(rln)) };
+    //let rln = RLN::new(tree_height, input_data);
+    //unsafe { *ctx = Box::into_raw(Box::new(rln)) };
```
Commented temporarily. Pay no attention to this.

richard-ramos commented on rln/src/ffi.rs:
```diff
@@ -59,8 +59,8 @@ pub extern "C" fn new_with_params(
     let vk_data = <&[u8]>::from(unsafe { &*vk_buffer });
-    let rln = RLN::new_with_params(tree_height, circom_data, zkey_data, vk_data);
-    unsafe { *ctx = Box::into_raw(Box::new(rln)) };
+    //let rln = RLN::new_with_params(tree_height, circom_data, zkey_data, vk_data);
+    //unsafe { *ctx = Box::into_raw(Box::new(rln)) };
```
Commented temporarily. Pay no attention to this.

richard-ramos commented on rln/src/ffi.rs:
```diff
@@ -198,6 +199,8 @@ pub extern "C" fn generate_rln_proof(
         false
     }
+    */
+    true
```
Commented temporarily. Pay no attention to this.

richard-ramos commented on rln/src/public.rs:
```diff
@@ -300,13 +304,14 @@ impl RLN<'_> {
 
-impl Default for RLN<'_> {
+/*
+impl Default for RLN {
     fn default() -> Self {
         let tree_height = TEST_TREE_HEIGHT;
         let buffer = Cursor::new(TEST_RESOURCES_FOLDER);
         Self::new(tree_height, buffer)
     }
-}
+}*/
```
Commented temporarily. Pay no attention to this.

richard-ramos commented on rln/src/protocol.rs:
```diff
@@ -374,13 +370,81 @@ pub enum ProofError {
     SynthesisError(#[from] SynthesisError),
 }
 
+fn calculate_witness_element<E: ark_ec::PairingEngine>(
+    witness: Vec<BigInt>,
+    sanity_check: bool,
+) -> Result<Vec<E::Fr>> {
+    use ark_ff::{FpParameters, PrimeField};
+    let modulus = <<E::Fr as PrimeField>::Params as FpParameters>::MODULUS;
+
+    // convert it to field elements
+    use num_traits::Signed;
+    let witness = witness
+        .into_iter()
+        .map(|w| {
+            let w = if w.sign() == num_bigint::Sign::Minus {
+                // Need to negate the witness element if negative
+                modulus.into() - w.abs().to_biguint().unwrap()
+            } else {
+                w.to_biguint().unwrap()
+            };
+            E::Fr::from(w)
+        })
+        .collect::<Vec<_>>();
+
+    Ok(witness)
+}
```
This function was extracted from [ark-circom](https://github.com/gakonst/ark-circom/blob/master/src/witness/witness_calculator.rs#L258-L286), It's main difference is that it will receive the witness instead of calculating it inside the function.

richard-ramos commented on rln/src/protocol.rs:
```diff
@@ -374,13 +370,81 @@ pub enum ProofError {
+    //let now = Instant::now();
+
+    // TODO: this function fails
+    let proof = create_proof_with_reduction_and_matrices::<_, CircomReduction>(
```
This function is failing with this error:
```
rln_bg.js:463 panicked at 'The global thread pool has not been initialized.: ThreadPoolBuildError { kind: IOError(Error { kind: Unsupported, message: "operation not supported on this platform" }) }', .cargo/registry/src/github.com-1ecc6299db9ec823/rayon-core-1.9.3/src/registry.rs:170:10
```

richard-ramos replied:
Not failing anymore ;)

richard-ramos commented:
How to compile zerokit for wasm and see example code:
1. Make sure you have nodejs installed and the `build-essential` package if using ubuntu.
2. Install wasm-pack
```
curl https://rustwasm.github.io/wasm-pack/installer/init.sh -sSf | sh
```
3. Compile zerokit for `wasm32-unknown-unknown`:
```
cd rln
wasm-pack build --release
```
4. Launch example app
```
cd www
npm install
npm start
```
5. Browse http://localhost:8080 and open the developer tools to see console logs

richard-ramos commented:
@s1fr0 , @staheri14 
I wasn't able to make zerokit load the wasm file (I believe it's due to reasons explained in this issue: https://github.com/gakonst/ark-circom/issues/4), so I ended up doing the calculation on the JS side, hence why I have two functions for generating the proof instead of one (to obtain the json input to pass to the witness calculator, and another function that receives the calculated witness).

I'm currently having trouble with `create_proof_with_reduction_and_matrices` because it tries to use threads even after disabling the `parallel` feature. (I'm not sure what's going on here). The error I'm receiving is 
```
panicked at 'The global thread pool has not been initialized.: ThreadPoolBuildError { kind: IOError(Error { kind: Unsupported, message: "operation not supported on this platform" }) }
```

I'll try cloning arkworks-rs/groth16 and adding some logs to that function to try to find out what's going on.

richard-ramos commented on rln/www/index.js:
```diff
@@ -0,0 +1,108 @@
+
+    function serializeMessage(uint8Msg, memIndex, epoch, idKey) {
+        // calculate message length
+        const msgLen = Buffer.allocUnsafe(8);
+        msgLen.writeUIntLE(uint8Msg.length, 0, 8);
+
+        // Converting index to LE bytes
+        const memIndexBytes = Buffer.allocUnsafe(8)
+        memIndexBytes.writeUIntLE(memIndex, 0, 8);
```
@fryorcraken I had to use a buffer here because i need the function `writeUintLE`

richard-ramos commented on rln/www/index.js:
```diff
@@ -0,0 +1,108 @@
+        return _concatenate(Uint8Array, idKey, memIndexBytes, epoch, msgLen, uint8Msg )
+    }
+
+const verifKeyUint8Array = _base64ToArrayBuffer(files.verification_key);
+const zkeyUint8Array = _base64ToArrayBuffer(files.zkey);
+const circomUint8Array = _base64ToArrayBuffer(files.rln_wasm);
```
Should a dev be able to specify the depth and the zkey and verification key to use? or should these be embedded in the code?

richard-ramos commented:
Good news! I managed to generate proofs now! I had to make a change in https://github.com/vacp2p/ark-circom/commit/dcde9ba5f1528ae1af649843cfe3241ecc4cb304 so the `parallels` feature can be disabled, and this allows generating the proof in the browser without running into the error:
```
panicked at 'The global thread pool has not been initialized.: ThreadPoolBuildError { kind: IOError(Error { kind: Unsupported, message: "operation not supported on this platform" }) }
```
However, the proof generation is very slow. It takes ~9s. Perhaps there's a way to enable threads on wasm?
For now, I'm happy that I was able to move past this blocker, so I can continue with verifying the proof, extracting the JS code to a separate repository and reverting a lot of the changes I made that are unrelated to the task of running zerokit on the browser.

richard-ramos commented:
I just noticed that if I don't open the devtools console immediatly, the proof generations is faster, i.e:
```
proof_gen_timer: 2492.31396484375 ms
```
~2.5s 

:thinking:

richard-ramos commented:
I think the code is now ready for a review. Please do not focus on the javascript code, as it is done only to see interop between js and rust: it is going to be removed from the repository, and is kept there so PR reviewers can see/test the wasm output. 
Please check previous comments regarding some changes done in rln/

richard-ramos commented:
I will close the PR as I learned today that compiling zerokit to wasm is not needed.
It was a nice learning exercise tho :-/

oskarth commented:
@richard-ramos First of all, amazing work on getting WASM support working!


I think there was some miscommunication here. The first/easiest thing to do here was to use existing JS RLN libraries, like zk-kit and other things from PSE. This seems like a riskier and harder path, but if it works that's great. Sanaz and Franck have some more context from PM call(s) about this too.

Regardless of if this is this the main path we choose or not we should definitely try to get it merged, so re-opening this. Will have a look at the PR a bit later.

oskarth reviewed (commented):
Trying this out, having some problems getting it to run.

```
    Compile zerokit for wasm32-unknown-unknown:

cd rln
wasm-pack build --release
```

I assume this should be `rln-wasm`? When I run in `rln` I get:

```
error: the wasm32-unknown-unknown target is not supported by default, you may need to enable the "js" feature. For more information see: https://docs.rs/getrandom/#webassembly-support
```

But when I run in `rln-wasm` the `wasm-pack` command succeeds.

`npm install` also works. However, when I run `npm start` I get:

```
ERROR in ../pkg/rln_wasm_bg.wasm
Import "__wbg_BigInt_1fab4952b6c4a499" from "./rln_wasm_bg.js" with Non-JS-compatible Func Sigurature (i64 as parameter) can only be used for direct wasm to wasm dependencies
 @ ../pkg/rln_wasm.js
 @ ./index.js
 @ ./bootstrap.js

ERROR in ../pkg/rln_wasm_bg.wasm
Import "__wbg_BigInt_d0c7d465bfa30d3b" from "./rln_wasm_bg.js" with Non-JS-compatible Func Sigurature (i64 as parameter) can only be used for direct wasm to wasm dependencies
 @ ../pkg/rln_wasm.js
 @ ./index.js
 @ ./bootstrap.js
ℹ ｢wdm｣: Failed to compile.
```

Any idea what could be wrong?

Env:

```
# Running on MBP M1
node --version: v18.9.0
wasm-pack --version: wasm-pack 0.10.3
rustc --version: rustc 1.63.0 (4b91a6ea7 2022-08-08)
npm --version: 8.19.1
```

oskarth commented:
Few general comments:

1) Can we target `master` on this instead of release branch? And resolve conflicts
2) Can we add a simple README in crate with instructions like you wrote in OP, and then a brief sentence in RLN crate about WASM support existing in separate crate?

oskarth commented on rln-wasm/src/lib.rs:
```diff
@@ -0,0 +1,128 @@
+    }
+
+    false
+}
```
Can we add some tests to this module? Super basic one

oskarth commented on rln-wasm/www/README.md:
```diff
@@ -0,0 +1,20 @@
+# RLN WASM EXAMPLE APP
```
Oh I see there's a README here already! I missed it because I expected it to be in the rln-wasm dir

richard-ramos replied:
This example app will also be removed, and will exist in `https://github.com/waku-org/js-rln/`  :)

oskarth commented on rln/Cargo.toml:
```diff
@@ -6,18 +6,21 @@ edition = "2021"
 ark-serialize = { version = "0.3.0", default-features = false }
-ark-circom = { git = "https://github.com/gakonst/ark-circom", features = ["circom-2"] }
-wasmer = { version = "2.0" }
+ark-circom = { git = "https://github.com/vacp2p/ark-circom", branch = "wasm", default-features = false, features = ["circom-2"] }
```
Can we upstream this as a PR to ark-circom too? Are all things in https://github.com/gakonst/ark-circom/issues/4 supported?

I think even if just partial support for our case it'd be a welcome PR!

oskarth replied:
Ah, I see in https://github.com/vacp2p/ark-circom/commit/0e587145cb05e08b2d1a01509eb578670088eb2f that it requires disabling `parallels`. I wonder if it is possible to get the feature flag to be disabled when `wasm` flag is enabled? Perhaps in a conditional package list or something, WDYT?

Also any idea why `parallels` doesn't work?

richard-ramos replied:
> any idea why `parallels` doesn't work?

While running with that feature enabled, the following error was generated:
```
The global thread pool has not been initialized.: ThreadPoolBuildError { kind: IOError(Error { kind: Unsupported, message: "operation not supported on this platform"
```

I was looking at https://github.com/GoogleChromeLabs/wasm-bindgen-rayon , which should enable concurrency using js web workers, but I feel like the effort of adding multithreading to zerokit-wasm should be done in a separate PR.

richard-ramos commented:
> Trying this out, having some problems getting it to run.
> 
> ```
>     Compile zerokit for wasm32-unknown-unknown:
> 
> cd rln
> wasm-pack build --release
> ```
> 
> I assume this should be `rln-wasm`? When I run in `rln` I get:
> 
> ```
> error: the wasm32-unknown-unknown target is not supported by default, you may need to enable the "js" feature. For more information see: https://docs.rs/getrandom/#webassembly-support
> ```
> 
> But when I run in `rln-wasm` the `wasm-pack` command succeeds.
> 
> `npm install` also works. However, when I run `npm start` I get:
> 
> ```
> ERROR in ../pkg/rln_wasm_bg.wasm
> Import "__wbg_BigInt_1fab4952b6c4a499" from "./rln_wasm_bg.js" with Non-JS-compatible Func Sigurature (i64 as parameter) can only be used for direct wasm to wasm dependencies
>  @ ../pkg/rln_wasm.js
>  @ ./index.js
>  @ ./bootstrap.js
> 
> ERROR in ../pkg/rln_wasm_bg.wasm
> Import "__wbg_BigInt_d0c7d465bfa30d3b" from "./rln_wasm_bg.js" with Non-JS-compatible Func Sigurature (i64 as parameter) can only be used for direct wasm to wasm dependencies
>  @ ../pkg/rln_wasm.js
>  @ ./index.js
>  @ ./bootstrap.js
> ℹ ｢wdm｣: Failed to compile.
> ```
> 
> Any idea what could be wrong?
> 
> Env:
> 
> ```
> # Running on MBP M1
> node --version: v18.9.0
> wasm-pack --version: wasm-pack 0.10.3
> rustc --version: rustc 1.63.0 (4b91a6ea7 2022-08-08)
> npm --version: 8.19.1
> ```

Please try again with the latest commit. I updated some js dependency versions, so it's necessary to do a `npm i` in `www` to download them.

oskarth commented:
Works now! I see the proof verification in the console :)

richard-ramos commented:
I added a couple of tests, and removed the example app since it's supposed to live in js-rln anyways. The only task pending is creating PRs for ark-circom

oskarth commented on rln-wasm/README.md:
````diff
@@ -0,0 +1,20 @@
+```
+cd rln-wasm
+wasm-pack build --release
+```
````
Maybe add a pointer to how https://github.com/waku-org/js-rln/? Maybe this belongs more in that repo, but looking at README not obvious to me how to run something similar to the example app that existed here before

oskarth reviewed (approved):
LGTM! Feel free to merge, can do ark-circom upstream update in separate PR
